server/src/main/java/com/hedera/block/server/consumer/LiveStreamObserverImpl.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Hedera Block Node
+ *
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.block.server.consumer;
+
+import com.hedera.block.protos.BlockStreamServiceGrpcProto;
+import com.hedera.block.server.mediator.StreamMediator;
+import io.grpc.stub.StreamObserver;
+
+import java.util.logging.Logger;
+
+/**
+ * The LiveStreamObserverImpl class implements the LiveStreamObserver interface to pass blocks to the downstream consumer
+ * via the notify method and manage the bidirectional stream to the consumer via the onNext, onError, and onCompleted methods.
+ *
+ */
+public class LiveStreamObserverImpl implements LiveStreamObserver<BlockStreamServiceGrpcProto.Block, BlockStreamServiceGrpcProto.BlockResponse> {
+
+    private final Logger LOGGER = Logger.getLogger(getClass().getName());
+
+    private final StreamMediator<BlockStreamServiceGrpcProto.Block, BlockStreamServiceGrpcProto.BlockResponse> mediator;
+    private final StreamObserver<BlockStreamServiceGrpcProto.Block> responseStreamObserver;
+
+    private long consumerLivenessMillis;
+    private long producerLivenessMillis;
+    private final long timeoutThresholdMillis;
+
+    /**
+     * Constructor for the LiveStreamObserverImpl class.
+     *
+     * @param mediator the mediator
+     * @param responseStreamObserver the response stream observer
+     *
+     */
+    public LiveStreamObserverImpl(final long timeoutThresholdMillis,
+            StreamMediator<BlockStreamServiceGrpcProto.Block, BlockStreamServiceGrpcProto.BlockResponse> mediator,
+            StreamObserver<BlockStreamServiceGrpcProto.Block> responseStreamObserver) {
+
+        this.mediator = mediator;
+        this.responseStreamObserver = responseStreamObserver;
+
+        this.timeoutThresholdMillis = timeoutThresholdMillis;
+        this.consumerLivenessMillis = System.currentTimeMillis();
+        this.producerLivenessMillis = System.currentTimeMillis();
+    }
+
+    /**
+     * Pass the block to the observer provided by Helidon
+     *
+     * @param block - the block to be passed to the observer
+     */
+    @Override
+    public void notify(final BlockStreamServiceGrpcProto.Block block) {
+
+        if (System.currentTimeMillis() - this.consumerLivenessMillis > timeoutThresholdMillis) {
```

Comment:
Similar here, it might be wise to use a Clock object (provided in the constructor) to get the current time here so we can manipulate the effective time for unit testing.

Reply:
fixed